linter/duplicate_routes/internal/duplicateroutes/duplicateroutes.go

```diff
@@ -0,0 +1,244 @@
+)
+
+// Finding represents a duplicated route within a protobuf file.
+type Finding struct {
+	File  string
+	Route string
+	Lines []int
+}
+
+// FileSystem encapsulates the filesystem interactions required by the scanner.
+type FileSystem interface {
+	Stat(string) (fs.FileInfo, error)
+	WalkDir(string, fs.WalkDirFunc) error
+	ReadFile(string) ([]byte, error)
+}
+
+type osFS struct{}
+
+func (osFS) Stat(name string) (fs.FileInfo, error) {
+	return os.Stat(name)
+}
+
+func (osFS) WalkDir(root string, fn fs.WalkDirFunc) error {
+	return filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		return fn(path, d, err)
+	})
+}
+
+func (osFS) ReadFile(name string) ([]byte, error) {
+	return os.ReadFile(name)
+}
+
+// Scan walks the provided paths, recursively searching for duplicate google.api.http routes.
+// When paths is empty, the current working directory is used.
+func Scan(paths []string) ([]Finding, error) {
+	return scanWithFilesystem(osFS{}, paths)
+}
+
+// ScanFS performs the scan against an arbitrary fs.FS implementation.
+// When roots is empty, the scan starts from ".".
+func ScanFS(fsys fs.FS, roots []string) ([]Finding, error) {
+	return scanWithFilesystem(fsAdapter{fsys: fsys}, roots)
+}
+
+func scanWithFilesystem(fsys FileSystem, paths []string) ([]Finding, error) {
+	if len(paths) == 0 {
+		paths = []string{"."}
+	}
+
+	aggregated := make(map[string]map[string][]int)
+	for _, root := range paths {
+		if err := processPath(fsys, root, aggregated); err != nil {
+			return nil, err
+		}
+	}
+
+	return finalizeFindings(aggregated), nil
+}
+
+func processPath(fsys FileSystem, root string, aggregated map[string]map[string][]int) error {
+	if root == "" {
+		root = "."
+	}
+
+	info, err := fsys.Stat(root)
+	if err != nil {
+		return err
+	}
+
+	if !info.IsDir() {
+		return processFile(fsys, root, aggregated)
+	}
+
+	return fsys.WalkDir(root, func(path string, d fs.DirEntry, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+		if d.IsDir() {
+			return nil
+		}
+		if filepath.Ext(path) != ".proto" {
+			return nil
+		}
+		return processFile(fsys, path, aggregated)
+	})
+}
+
+func processFile(fsys FileSystem, path string, aggregated map[string]map[string][]int) error {
+	if filepath.Ext(path) != ".proto" {
+		return nil
+	}
+
+	duplicates, err := findDuplicates(fsys, path)
+	if err != nil {
+		return err
+	}
+	if len(duplicates) == 0 {
+		return nil
+	}
+
+	if _, ok := aggregated[path]; !ok {
+		aggregated[path] = make(map[string][]int)
+	}
+
+	for route, lines := range duplicates {
+		aggregated[path][route] = append(aggregated[path][route], lines...)
+	}
+
+	return nil
+}
+
+func findDuplicates(fsys FileSystem, path string) (map[string][]int, error) {
+	data, err := fsys.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	lines := strings.Split(string(data), "\n")
+	occur := make(map[string][]int)
+
+	for index := 0; index < len(lines); index++ {
```

Comment:
To avoid loading the whole file in memory there's already the standard `Scanner` type allowing to read line by line, for example:


```golang
file, err := f.Open(path)
if err != nil {
	return nil, err
}
scanner := bufio.NewScanner(file)
for scanner.Scan() {
	line := scanner.Text()
	...
}
if err := scanner.Err(); err != nil {
	return nil, err
}
```